### ISSUE TITLE
blob: add an As method to Writer

### ISSUE THREAD

**Issue by afking** [google/go-cloud#2520]
To make reads/writes behave more like io.Copy I would like to access `ReaderFrom` methods on the underlying Writer type. As() methods seems like the way to go as this can be done on the Reader type already.

### Proposed Change

```
func (w *Writer) As(v interface{}) bool {
    ...
}
```

**Comment by shantuo** [comment]
Hi @afking , which provider are you using? I'm asking because most providers' writer types don't implement `ReaderFrom`.

**Comment by afking** [comment]
Hello, it's a custom writer for stfp transfers. It doesn't really fit the blob model but I've found it super useful to have the abstraction for testing  etc. It does some connection multiplexing which I'm loosing by using blob.Writer: https://godoc.org/github.com/pkg/sftp#File.ReadFrom

**Comment by vangent** [comment]
Hi @afking -- 

TL;DR: Use the `BeforeWrite` field of [WriterOptions](https://godoc.org/gocloud.dev/blob#WriterOptions). You'll provide a function that will get called before the first actual write, and you can do the `As` there (see the example in the godoc).

More details:

The reason we do it this way rather than as a direct `Writer.As` function is that sometimes the provider-specific type for writing is not created at the point when the `blob.Writer` is returned to the user.

Even more details:

The `blob.driver.NewTypedWriter` function takes a required `ContentType` argument; if the user doesn't supply one to `blob.NewWriter`, the blob portable type will [intercept](https://github.com/google/go-cloud/blob/master/blob/blob.go#L791-L796) the first few bytes, use them to determine a content type, and then lazily call `driver.NewTypedWriter` later.

Some providers also don't have all of the information they need until just before the write happens.

**Comment by vangent** [comment]
(I closed this because I'm pretty sure `WriterOptions.BeforeWrite` addresses this issue; please confirm/deny).

**Comment by afking** [comment]
@vangent thanks for the clarification! I think it would work.

**Comment by afking** [comment]
@vangent although it is possible with `BeforeWrite` method it feels a little odd as the transfer will be done during creation:
```
w, err := bkt.NewWriter(..., &blob.WriterOptions{
    BeforeFunc: func(asFunc func(interface{}) bool) error {
            var wt *custom.WriterType
            if asFunc(wt) {
                _, err := wt.ReadFrom(r)
               return err
           }
           return nil
    }
})
w.Close()
```
The Reader has a BeforeRead method and As. The use case is probably small but for symmetry could it be added to the the writer? Then:

```
var wt *custom.WriterType
if w.As(wt) {
     retrun wt.ReadFrom(r)
}
return io.Copy(w, r)
```

**Comment by vangent** [comment]
Sorry, I don't understand your code snippet. What are you actually trying to do?

`As` functions (both the top-level ones and the `BeforeRead`/`BeforeWrite` callback ones) are used to access the underlying provider types (for example, you can see the types that will work for `gcsblob` [here](https://godoc.org/gocloud.dev/blob/gcsblob#hdr-As)). They won't convert to arbitrary types like `custom.WriterType`.

**Comment by afking** [comment]
The `custom.WriterType` is type from a non public blob driver that implements it. The code snippets above was trying to show how to write a reader `r io.Reader` to a `blob.Writer` by using in my case the `io.ReaderFrom` interface that the `custom.WriterType` implements.

A gcs storage example might be trying to get `Attrs()` https://godoc.org/cloud.google.com/go/storage#Writer.Attrs from a writer after successfully writing.
```
io.Copy(w, r)
var wt *storage.Writer
if w.As(wt) {
     fmt.Println(wt.Attrs())
}
```

**Comment by vangent** [comment]
I think I see. Basically, you don't want to use the `blob.Writer` at all (i.e., you will never write data to it) for a particular driver, but short-circuit it.

I think to do that you should use the top-level `As` on `blob.Bucket`, something like this:

```
var c *sftp.Client // or whatever type your driver exposes for Bucket.As
if bucket.As(&c) {
  // Do the short-circuit write.
  f, _ := c.Open(key)
  f.ReadFrom(r)
} else {
  w, _ := bucket.NewWriter(key, nil)
  io.Copy(w, r)
}

This makes it explicit that you are not using `blob.Writer`.  


For the gcs storage example, you would do something like this:

```
var sw *storage.Writer
fn := func(asFunc func(interface{}) bool) error {
  asFunc(&sw)
  return nil
}
w, _ := bucket.NewWriter(..., &blob.WriterOptions{BeforeWrite: fn})
io.Copy(w, r)
if err := w.Close(); err == nil && sw != nil {
  fmt.Printf("%v\n", sw.Attrs())
}
```

Hope that helps.

**Comment by afking** [comment]
Yes that makes sense! Thanks for taking the time to explain.
